File: modules/twake-segmentation/android/src/main/java/com/linagora/twakevisio/segmentation/EffectFrameProcessor.kt
package com.linagora.twakevisio.segmentation

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.Rect
import android.os.SystemClock
import android.util.Log
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.segmentation.Segmentation
import com.google.mlkit.vision.segmentation.selfie.SelfieSegmenterOptions
import org.webrtc.VideoFrame
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicReference

/** Ce que l'utilisateur a choisi. `None` est un état, pas une absence. */
sealed interface BackgroundEffect {
  data object None : BackgroundEffect
  data object Blur : BackgroundEffect
  data class Image(val bitmap: Bitmap) : BackgroundEffect
}

/**
 * Le cœur de l'étape 4 : segmenter, puis composer.
 *
 * **La segmentation tourne à BASSE DÉFINITION et le masque est agrandi.** C'est
 * la décision du propriétaire, et la mesure de l'étape 1 la rendait obligatoire :
 * 17,96 ms à 480p mais 26,04 ms à 720p, sur 33,33 ms disponibles par image — et
 * à 720p une image sur vingt dépassait le budget à elle seule. Segmenter en
 * pleine définition ne laissait la place à rien d'autre.
 *
 * Le coût de ce choix est visible au CONTOUR, pas sur le sujet : la personne
 * reste nette, seule la frontière est un peu moins précise. C'est le compromis
 * que fait aussi le client web de la DINUM.
 *
 * `None` court-circuite tout et rend l'image d'origine. Le décorateur reste donc
 * en place en permanence, et activer un effet **ne reconstruit pas la piste** —
 * donc aucune renégociation au milieu d'une réunion.
 */
class EffectFrameProcessor : SegmentingCapturer.FrameProcessor {

  private val effect = AtomicReference<BackgroundEffect>(BackgroundEffect.None)

  // STREAM_MODE : le mode d'un flux, qui réutilise l'état d'une image à
  // l'autre. C'est celui qui a été mesuré à l'étape 1.
  private val segmenter = Segmentation.getClient(
    SelfieSegmenterOptions.Builder()
      .setDetectorMode(SelfieSegmenterOptions.STREAM_MODE)
      .build(),
  )

  // Réutilisés d'une image à l'autre. Allouer un bitmap par image ferait passer
  // le ramasse-miettes plus souvent que la caméra ne produit d'images.
  private var maskBitmap: Bitmap? = null
  private var frameCount = 0L
  private var totalNanos = 0L

  fun setEffect(next: BackgroundEffect) {
    effect.set(next)
  }

  fun dispose() {
    segmenter.close()
    maskBitmap?.recycle()
    maskBitmap = null
  }

  override fun process(frame: VideoFrame): VideoFrame {
    val current = effect.get()
    // LE chemin rapide, et la raison pour laquelle le décorateur peut rester
    // branché en permanence : aucun effet, aucun travail.
    if (current is BackgroundEffect.None) return frame

    val started = SystemClock.elapsedRealtimeNanos()
    return try {
      val i420 = frame.buffer.toI420() ?: return frame
      try {
        val output = composite(i420, current, frame.rotation) ?: return frame
        VideoFrame(output, frame.rotation, frame.timestampNs)
      } finally {
        i420.release()
      }
    } catch (error: Exception) {
      // Une image ratée ne doit JAMAIS interrompre l'appel : on publie
      // l'originale. Le journal dit ce qui s'est passé ; l'interlocuteur voit
      // une image sans effet plutôt qu'un écran noir.
      Log.w(TAG, "composition échouée, image d'origine publiée", error)
      frame
    } finally {
      frameCount += 1
      totalNanos += SystemClock.elapsedRealtimeNanos() - started
      // Une moyenne toutes les 120 images — quatre secondes à 30 i/s. Assez
      // rare pour ne pas peser, assez fréquent pour voir une dérive.
      if (frameCount % 120 == 0L) {
        Log.i(TAG, "coût moyen ${"%.2f".format(totalNanos / frameCount / 1_000_000.0)} ms " +
          "sur $frameCount images")
      }
    }
  }

  private fun composite(
    i420: VideoFrame.I420Buffer,
    current: BackgroundEffect,
    rotation: Int,
  ): VideoFrame.Buffer? {
    val width = i420.width
    val height = i420.height

    // 1. L'image réduite pour MLKit. `SEGMENTATION_WIDTH` est la définition à
    //    laquelle l'étape 1 a mesuré 17,96 ms.
    val scale = SEGMENTATION_WIDTH.toDouble() / width
    val segWidth = SEGMENTATION_WIDTH
    val segHeight = (height * scale).toInt().coerceAtLeast(1)
    val small = FrameConversion.toScaledBitmap(i420, segWidth, segHeight)

    // 2. Le masque, à la définition de l'entrée réduite.
    val mask = Tasks.await(
      segmenter.process(InputImage.fromBitmap(small, 0)),
      SEGMENTATION_TIMEOUT_MS,
      TimeUnit.MILLISECONDS,
    )
    val confidences = mask.buffer
    confidences.rewind()

    // Un bitmap d'alpha : 255 là où la personne est, 0 ailleurs. C'est lui
    // qu'on agrandit — pas le masque flottant, qu'il faudrait interpoler à la
    // main.
    val alphaPixels = IntArray(mask.width * mask.height)
    for (index in alphaPixels.indices) {
      val confidence = confidences.float
      // La confiance brute NE PEUT PAS servir d'alpha telle quelle, et c'est le
      // défaut qui rendait le sujet translucide : MLKit rend souvent 0,6 ou 0,8
      // à l'INTÉRIEUR de la personne, pas 1. « Probablement une personne »
      // devenait donc « à moitié transparente ».
      //
      // On durcit : opaque au-dessus du seuil haut, transparent en dessous du
      // seuil bas, et une rampe entre les deux. La rampe est ce qui garde un
      // bord doux plutôt qu'un découpage à l'emporte-pièce — c'est elle qui
      // évite l'aspect « autocollant » d'un masque binaire.
      val alpha = when {
        confidence >= OPAQUE_ABOVE -> 255
        confidence <= CLEAR_BELOW -> 0
        else -> (((confidence - CLEAR_BELOW) / (OPAQUE_ABOVE - CLEAR_BELOW)) * 255).toInt()
      }
      alphaPixels[index] = alpha.coerceIn(0, 255) shl 24
    }
    val maskSmall = Bitmap.createBitmap(alphaPixels, mask.width, mask.height, Bitmap.Config.ARGB_8888)

    // 3. L'image PLEINE, celle qu'on publie.
    val full = FrameConversion.toScaledBitmap(i420, width, height)

    // 4. Le fond : flou de l'image elle-même, ou l'image choisie.
    val background = when (current) {
      is BackgroundEffect.Blur -> cheapBlur(full)
      // PIVOTÉ, et c'est le défaut que le propriétaire a vu : la composition se
      // fait dans l'orientation NATIVE du capteur — 640x480 paysage — puis
      // `VideoFrame(…, rotation, …)` fait pivoter le tout à l'affichage. Un
      // fond dessiné tel quel arrivait donc couché. Le pivoter à contresens ici
      // le redresse une fois la rotation d'affichage appliquée.
      //
      // La personne, elle, n'a jamais eu ce problème : elle vient de la même
      // image paysage et subit la même rotation.
      is BackgroundEffect.Image -> orientToFrame(current.bitmap, width, height, rotation)
      BackgroundEffect.None -> return null
    }

    // 5. La composition. `DST_IN` garde la personne là où le masque est opaque ;
    //    le fond est dessiné d'abord, la personne découpée par-dessus. Tout est
    //    fait par `Canvas`, donc en natif — c'est ce qui rend le coût tenable.
    val result = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
    val canvas = Canvas(result)
    val destination = Rect(0, 0, width, height)
    canvas.drawBitmap(background, null, destination, null)

    val cutout = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
    val cutoutCanvas = Canvas(cutout)
    cutoutCanvas.drawBitmap(full, 0f, 0f, null)
    val maskPaint = Paint().apply { xfermode = PorterDuffXfermode(PorterDuff.Mode.DST_IN) }
    cutoutCanvas.drawBitmap(maskSmall, null, destination, maskPaint)
    canvas.drawBitmap(cutout, 0f, 0f, null)

    val buffer = FrameConversion.toI420Buffer(result)

    small.recycle()
    maskSmall.recycle()
    full.recycle()
    cutout.recycle()
    result.recycle()
    // `orientToFrame` et `cheapBlur` rendent tous deux une COPIE : on relâche
    // toujours, et jamais le bitmap d'origine gardé en cache.
    background.recycle()

    return buffer
  }

  /**
   * Redresse et cadre un fond pour la composition.
   *
   * Deux opérations, dans cet ordre : pivoter de `-rotation` pour annuler la
   * rotation que l'affichage appliquera, puis RECADRER en conservant les
   * proportions. Un `drawBitmap` sur un rectangle plein étirerait l'image —
   * c'est ce que faisait la version précédente, et sur une vignette 107x60
   * étirée en 640x480 le résultat était méconnaissable.
   */
  private fun orientToFrame(source: Bitmap, width: Int, height: Int, rotation: Int): Bitmap {

    // Recadrage « couvrir », sur le rapport de l'image AFFICHÉE — pas celui du
    // tampon.
    //
    // Le tampon est en 640x480 paysage, mais la rotation de 90° le présente en
    // 480x640 portrait. Découper sur 640/480 revenait à cadrer pour un écran
    // couché : on prenait une bande horizontale du fond redressé, d'où l'effet
    // de zoom que le propriétaire a signalé.
    //
    // Une rotation de 90 ou 270 ÉCHANGE les côtés vus. C'est ce rapport-là qu'il
    // faut viser.
    val swapped = rotation % 180 != 0
    val displayWidth = if (swapped) height else width
    val displayHeight = if (swapped) width else height
    val targetRatio = displayWidth.toFloat() / displayHeight
    val sourceRatio = source.width.toFloat() / source.height
    val cropWidth: Int
    val cropHeight: Int
    if (sourceRatio > targetRatio) {
      cropHeight = source.height
      cropWidth = (cropHeight * targetRatio).toInt().coerceAtMost(source.width)
    } else {
      cropWidth = source.width
      cropHeight = (cropWidth / targetRatio).toInt().coerceAtMost(source.height)
    }
    // On découpe d'ABORD, à la bonne proportion d'affichage, PUIS on redresse.
    // L'ordre inverse — redresser puis découper — cadrait sur une image déjà
    // pivotée et reprenait le mauvais rapport.
    val cropped = Bitmap.createBitmap(
      source,
      (source.width - cropWidth) / 2,
      (source.height - cropHeight) / 2,
      cropWidth,
      cropHeight,
    )
    val matrix = Matrix().apply { postRotate(-rotation.toFloat()) }
    val upright = if (rotation % 360 == 0) cropped else Bitmap.createBitmap(
      cropped, 0, 0, cropped.width, cropped.height, matrix, true,
    )
    val scaled = Bitmap.createScaledBitmap(upright, width, height, true)
    if (upright !== cropped) upright.recycle()
    if (cropped !== source) cropped.recycle()
    return scaled
  }

  /**
   * Un flou par réductions et agrandissements SUCCESSIFS.
   *
   * La version précédente réduisait d'un coup d'un facteur 12 puis réétirait :
   * le résultat n'était pas un flou, c'étaient de gros carrés. Le propriétaire
   * l'a vu immédiatement sur appareil, et le commentaire qui promettait « un
   * flou convaincant » était faux.
   *
   * La correction tient à l'arithmétique du filtrage bilinéaire : il n'échantillonne
   * que les QUATRE voisins immédiats. Réduire d'un facteur 12 en une passe jette
   * donc 143 pixels sur 144 sans les moyenner — d'où les blocs. Enchaîner des
   * réductions de moitié fait entrer chaque pixel dans la moyenne à chaque
   * passe, ce qui approche un vrai noyau gaussien.
   *
   * Quatre allers-retours de moitié coûtent quelques opérations natives, là où
   * un noyau gaussien à la main coûterait la surface de l'image par le carré du
   * rayon.
   */
  private fun cheapBlur(source: Bitmap): Bitmap {
    var current = source
    var owned = false
    // Descente : quatre passes de moitié, soit un seizième de la définition.
    repeat(BLUR_PASSES) {
      val next = Bitmap.createScaledBitmap(
        current,
        (current.width / 2).coerceAtLeast(1),
        (current.height / 2).coerceAtLeast(1),
        true,
      )
      if (owned) current.recycle()
      current = next
      owned = true
    }
    // Remontée : on repasse par les mêmes paliers plutôt que d'étirer d'un
    // coup. Chaque palier remoyenne, et c'est ce qui efface les arêtes.
    repeat(BLUR_PASSES) {
      val next = Bitmap.createScaledBitmap(
        current,
        (current.width * 2).coerceAtMost(source.width),
        (current.height * 2).coerceAtMost(source.height),
        true,
      )
      current.recycle()
      current = next
    }
    return if (current.width == source.width && current.height == source.height) {
      current
    } else {
      val exact = Bitmap.createScaledBitmap(current, source.width, source.height, true)
      current.recycle()
      exact
    }
  }

  private companion object {
    const val TAG = "TwakeSegEffect"
    // La définition mesurée à l'étape 1 : 17,96 ms de médiane.
    const val SEGMENTATION_WIDTH = 640
    const val SEGMENTATION_TIMEOUT_MS = 200L
    // Quatre passes de moitié : un seizième de la définition, atteint par
    // moyennes successives plutôt que d'un seul saut.
    const val BLUR_PASSES = 4
    // Au-dessus : franchement la personne. En dessous : franchement le fond.
    // Entre les deux, une rampe de 0,25 de large, qui fait le bord doux.
    const val OPAQUE_ABOVE = 0.55f
    const val CLEAR_BELOW = 0.30f
  }
}

File: modules/twake-segmentation/ios/EffectFrameProcessor.swift
import CoreImage
import CoreImage.CIFilterBuiltins
import CoreVideo
import Foundation
import Vision
import WebRTC

/// Ce que l'utilisateur a choisi. `none` est un état, pas une absence.
enum TwakeBackgroundEffect {
  case none
  case blur
  case image(CIImage)
}

/**
 Le cœur du pendant iOS : segmenter, puis composer.

 **Rien de ce fichier ne ressemble à son homologue Android, et c'est le point.**
 Là-bas, chaque image traverse deux conversions I420 ↔ ARGB écrites à la main,
 qui pèsent à elles seules le tiers du budget. Ici, `RTCVideoFrame` porte déjà
 un `CVPixelBuffer`, Vision le prend tel quel, Core Image compose sur le GPU et
 rend dans un autre `CVPixelBuffer`. **Aucune boucle par pixel n'est écrite
 nulle part.**

 Ce n'est donc pas une transposition du code Android : c'est le même effet obtenu
 par le chemin que la plateforme rend naturel. Transposer aurait reproduit des
 conversions qu'iOS n'impose pas.

 La contrepartie est qu'aucun chiffre n'est encore mesuré ici. Le simulateur ne
 publie ni caméra ni micro — c'est écrit dans `AGENTS.md` —, donc la mesure
 demande un iPhone branché. Tant qu'elle n'a pas eu lieu, ce fichier compile et
 rien de plus : **on ne saura pas ce qu'il coûte avant de l'avoir mesuré sur
 appareil**, et il ne faut pas écrire ailleurs qu'il est rapide.
 */
final class EffectFrameProcessor {

  /// Rendu GPU. `CIContext` est cher à construire et sûr à réutiliser.
  private let ciContext = CIContext(options: [.cacheIntermediates: false])

  /// `.balanced` plutôt que `.accurate` : le mode précis vise la photo, pas les
  /// 33 ms d'une image de flux. `.fast` rend un masque nettement plus grossier.
  private let request: VNGeneratePersonSegmentationRequest = {
    let request = VNGeneratePersonSegmentationRequest()
    request.qualityLevel = .balanced
    request.outputPixelFormat = kCVPixelFormatType_OneComponent8
    return request
  }()

  private let lock = NSLock()
  private var effect: TwakeBackgroundEffect = .none

  /// Le masque conservé entre deux images, pour la même raison qu'Android :
  /// une silhouette ne bouge pas en 33 ms, et redécouvrir la même à chaque
  /// image est le poste le plus cher de la chaîne.
  private var retainedMask: CIImage?
  private var frameIndex: UInt64 = 0

  private var pool: CVPixelBufferPool?
  private var poolWidth = 0
  private var poolHeight = 0

  func setEffect(_ next: TwakeBackgroundEffect) {
    lock.lock()
    effect = next
    if case .none = next { retainedMask = nil }
    lock.unlock()
  }

  func process(_ frame: RTCVideoFrame) -> RTCVideoFrame {
    lock.lock()
    let current = effect
    lock.unlock()

    // LE chemin rapide, et la raison pour laquelle le délégué peut rester
    // branché en permanence : aucun effet, aucun travail.
    if case .none = current { return frame }

    guard let rtcBuffer = frame.buffer as? RTCCVPixelBuffer else { return frame }
    let pixelBuffer = rtcBuffer.pixelBuffer
    let source = CIImage(cvPixelBuffer: pixelBuffer)
    let extent = source.extent

    frameIndex &+= 1
    if retainedMask == nil || frameIndex % UInt64(Self.segmentEvery) == 0 {
      retainedMask = makeMask(from: pixelBuffer, fitting: extent)
    }
    guard let mask = retainedMask else { return frame }

    let background: CIImage
    switch current {
    case .blur:
      // `clampedToExtent` AVANT le flou : sans lui, le flou échantillonne du
      // transparent au-delà des bords et rend un liseré sombre tout autour de
      // l'image. Le défaut est discret au centre et net dans les coins.
      background = source
        .clampedToExtent()
        .applyingGaussianBlur(sigma: Self.blurSigma)
        .cropped(to: extent)
    case let .image(picture):
      background = Self.cover(picture, in: extent)
    case .none:
      return frame
    }

    let blend = CIFilter.blendWithMask()
    blend.inputImage = source
    blend.backgroundImage = background
    blend.maskImage = mask
    guard let composited = blend.outputImage else { return frame }

    guard let output = makePixelBuffer(width: Int(extent.width), height: Int(extent.height)) else {
      return frame
    }
    ciContext.render(composited, to: output)

    return RTCVideoFrame(
      buffer: RTCCVPixelBuffer(pixelBuffer: output),
      rotation: frame.rotation,
      timeStampNs: frame.timeStampNs,
    )
  }

  /**
   Le masque de silhouette, mis à l'échelle de l'image.

   Vision rend un masque à SA définition, pas à celle de l'entrée — souvent bien
   plus petit. Le composer sans l'agrandir découperait la personne dans un coin
   de l'image. C'est la même mise à l'échelle que fait `drawBitmap(mask, null,
   destination, …)` côté Android, ici portée par une transformation affine que
   Core Image applique sur le GPU.
   */
  private func makeMask(from pixelBuffer: CVPixelBuffer, fitting extent: CGRect) -> CIImage? {
    let handler = VNImageRequestHandler(cvPixelBuffer: pixelBuffer, options: [:])
    do {
      try handler.perform([request])
    } catch {
      // Une image ratée ne doit JAMAIS interrompre l'appel : on garde le masque
      // précédent, et à défaut l'appelant publiera l'image d'origine.
      NSLog("TwakeSegEffect: segmentation échouée — %@", error.localizedDescription)
      return retainedMask
    }
    guard let observation = request.results?.first else { return retainedMask }

    let raw = CIImage(cvPixelBuffer: observation.pixelBuffer)
    let scaleX = extent.width / raw.extent.width
    let scaleY = extent.height / raw.extent.height
    return raw.transformed(by: CGAffineTransform(scaleX: scaleX, y: scaleY))
  }

  /**
   Cadre une image de fond en « couvrir », sans jamais l'étirer.

   Le défaut que le propriétaire a vu sur Android était exactement celui-là :
   une image dessinée sur un rectangle plein est ÉTIRÉE, et un fond au mauvais
   rapport devient méconnaissable. On agrandit donc du facteur le plus grand des
   deux, puis on centre et on rogne.
   */
  private static func cover(_ picture: CIImage, in extent: CGRect) -> CIImage {
    let scale = max(extent.width / picture.extent.width, extent.height / picture.extent.height)
    let scaled = picture.transformed(by: CGAffineTransform(scaleX: scale, y: scale))
    let offsetX = extent.midX - scaled.extent.midX
    let offsetY = extent.midY - scaled.extent.midY
    return scaled
      .transformed(by: CGAffineTransform(translationX: offsetX, y: offsetY))
      .cropped(to: extent)
  }

  /**
   Un tampon de sortie tiré d'une RÉSERVE, jamais alloué à l'image.

   `CVPixelBufferCreate` par image ferait passer le ramasse-miettes plus souvent
   que la caméra ne produit d'images — c'est la même raison qui fait réutiliser
   les tampons côté Android, où elle a valu 3,5x sur la conversion d'entrée.
   */
  private func makePixelBuffer(width: Int, height: Int) -> CVPixelBuffer? {
    if pool == nil || poolWidth != width || poolHeight != height {
      let attributes: [String: Any] = [
        kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_420YpCbCr8BiPlanarFullRange,
        kCVPixelBufferWidthKey as String: width,
        kCVPixelBufferHeightKey as String: height,
        // Sans cette clé, Core Image ne peut pas écrire dans le tampon par le
        // GPU et retombe silencieusement sur un chemin logiciel.
        kCVPixelBufferIOSurfacePropertiesKey as String: [:] as CFDictionary,
      ]
      var created: CVPixelBufferPool?
      CVPixelBufferPoolCreate(kCFAllocatorDefault, nil, attributes as CFDictionary, &created)
      pool = created
      poolWidth = width
      poolHeight = height
    }
    guard let pool else { return nil }
    var buffer: CVPixelBuffer?
    CVPixelBufferPoolCreatePixelBuffer(kCFAllocatorDefault, pool, &buffer)
    return buffer
  }

  /// Une segmentation toutes les DEUX images, comme sur Android — où la mesure
  /// a montré que MLKit pesait 44 % du coût pour redécouvrir une silhouette
  /// immobile. Le chiffre iOS reste à mesurer sur appareil.
  private static let segmentEvery = 2

  /// Le rayon du flou. Core Image fait un VRAI gaussien sur le GPU, là où
  /// Android enchaîne des réductions de moitié faute de mieux.
  private static let blurSigma = 12.0
}
